theano/tensor/basic.py

```diff
@@ -6039,6 +6039,54 @@ def outer(x, y):
         y.dimshuffle('x', 0))
 
 
+def _matmul2(x, y):
+    """numpy.matmul() equivalent for 2D-or-more tensors only."""
+
+    # Make number of dimensions match, prepend broadcastable axes if needed.
+    newdims = abs(x.ndim - y.ndim)
+    if x.ndim < y.ndim:
+        x = x.dimshuffle(*(['x'] * newdims + list(range(x.ndim))))
+        dims = y.ndim
+    else:
+        y = y.dimshuffle(*(['x'] * newdims + list(range(y.ndim))))
+        dims = x.ndim
+
+    # Align the last axis of @x with the last second axis of @y
+    x = x.dimshuffle(*(list(range(dims)) + ['x']))
+    y = y.dimshuffle(*(list(range(dims - 2)) + ['x', dims - 2, dims - 1]))
+
+    # Compute the "elementwise" matrix product
+    return (x * y).sum(axis=-2)
```

Comment:
I found that the `BatchedDot` implementation looks awfully complicated :/

I'm thinking maybe just using it directly for ndim <= 3 and raising a `NotImplementedError` for ndim>3.
In my PR there I use a reshape trick to get matrices down to 3 dimensions, but it most likely doesn't play nice with broadcastable dimensions.